internal/proxysql/core.go

```diff
@@ -1,150 +1,231 @@
+	)
 
-	for {
-		p.coreLoop()
+	podInformer := factory.Core().V1().Pods().Informer()
 
-		time.Sleep(time.Duration(interval) * time.Second)
+	defer runtime.HandleCrash()
+
+	go factory.Start(stopper)
+
+	if !cache.WaitForCacheSync(stopper, podInformer.HasSynced) {
+		runtime.HandleError(fmt.Errorf("Timed out waiting for caches to sync"))
+		return
 	}
-}
 
-func (p *ProxySQL) coreLoop() {
-	pods, err := GetCorePods(p.settings)
+	_, err := podInformer.AddEventHandler(cache.ResourceEventHandlerFuncs{
+		AddFunc:    p.podAdded,
+		UpdateFunc: p.podUpdated,
+	})
 	if err != nil {
-		slog.Error("Failed to get pod info", slog.Any("error", err))
+		fmt.Println(err)
+	}
 
+	// block the main go routine from exiting
+	<-stopper
+}
+
+// This function is needed to do bootstrapping. At first I was using podUpdated to do adds, but we would never
+// get the first pod to come up. This function will only be useful on the first core pod to come up, the rest will
+// be handled via podUpdated.
+//
+// This feels a bit clumsy.
+func (p *ProxySQL) podAdded(object interface{}) {
+	pod, ok := object.(*v1.Pod)
+	if !ok {
 		return
 	}
 
-	if len(pods) == 0 {
-		slog.Error("No pods returned")
-
+	// if the new pod is not THIS pod, bail out of this function. the rest of this function should only apply
+	// to the first core pod to come up in the cluster.
+	if hostname, _ := os.Hostname(); pod.Name != hostname {
 		return
 	}
 
-	checksumFile := "/tmp/pods-cs.txt"
-	digest := calculateChecksum(pods)
+	// check if pod is already in the proxysql_servers table; this can happen when core pods add
+	// other core pods.
+	query := fmt.Sprintf("SELECT count(*) FROM proxysql_servers WHERE hostname = '%s'", pod.Status.PodIP)
+
+	var count int
 
-	// Read the previous checksum from the file
-	old, err := os.ReadFile(checksumFile)
+	err := p.conn.QueryRow(query).Scan(&count)
 	if err != nil {
-		old = []byte("")
+		slog.Error("Error in podAdded()", slog.Any("err", err))
 	}
 
-	// If nothing changes, we still run LOAD PROXYSQL SERVERS TO RUNTIME in order to accept any
-	// new pods that have joined the cluster
-	if string(old) == digest {
-		command := "LOAD PROXYSQL SERVERS TO RUNTIME"
+	if count > 0 {
+		return
+	}
 
-		_, err = p.conn.Exec(command)
-		if err != nil {
-			slog.Error("Command failed to execute", slog.String("command", command), slog.Any("error", err))
-		}
+	err = p.addPodToCluster(pod)
+	if err != nil {
+		slog.Error("Error in podAdded()", slog.Any("err", err))
+	}
+}
 
+// We aren't using podAdded here when other core pods exist because that function doesn't always get the PodIP,
+// and this one does. Using this function doesn't work when bootstrapping a cluster, because the pod has started
+// before the informer has started. In other words, the pod starts before the pod can detect itself joining the
+// cluster.
+//
+// Example pod (scaled up core-1, then scaled it back down):
+//
+//	OLD POD NAME	OLD POD IP			OLD STATUS	NEW POD NAME		NEW POD IP			NEW STATUS
```

Comment:
Is this just an issue with how Github is rendering this?

Reply:
what specifically, the spacing? i did that on purpose so i could read it easier

Reply:
Oh sorry I mean that the column names don't line up with the data below AFAICT